### DISCUSSION TITLE
RasPi logic levels

### DISCUSSION THREAD

**Discussion by IMHarris** [Dan-in-CA/SIP#286] [General]
Hi Dan-in-CA,

Looking at the [relay interface diagram](https://github.com/Dan-in-CA/SIP/wiki/Relay-interface), it shows 5v going to the shift register.  Given that the raspi operates on 3.3v logic levels, that can create some flaky behavior with the register (which I've experienced).  I'd suggest that the register be powered at 3.3v (RasPi pin 1).  My relays, powered at 5v, seem to work fine with a 3.3v logical signal output from the shift register.

This is great software and I'm planning on using it to run drip irrigation in my garden.  I'm also in the process of writing a flow plugin.  It's about 80% done and I'm looking forward to sharing it with this community in the next 6-8 weeks.

Ian

**Comment by Dan-in-CA** [comment]
Hi Ian,
Regarding the logic level issue, there is a newer port expander option using I2C.
It requires a plugin named pcf857x_plugin and you can read the documentation at:
https://gitlab.com/seventer/pcf857x_plugin_sip/-/blob/main/doc/pcf857x-doc.md

Although the  shift register interface does not require a software plugin, the I2C interface is easier to set up, less expensive, and more reliable.

A flow plugin would be a great addition to SIP. There was a flow plugin created some time ago but the author kept it as a separate fork and not part of the regular SIP plugin collection. You may find some of the code useful for your project. It can be found at:
https://github.com/DaveSprague/sip_plugins/tree/master/flow_sensors

Some of the modifications to the core SIP software that were required for that project have been incorporated into the main branch of SIP. That includes the option to display real time flow data on the home page timeline. Examples of how to use that feature can be found in the proto plugin that is shipped with SIP. 

I use drip for all my home irrigation. With the severe drought here in California there are restrictions on outdoor irrigation. The use of drip for food gardens has been exempt from the restrictions so far.

Looking forward to your flow plugin.

Dan

**Reply by IMHarris** [reply]
Thanks Dan,

Right now, my plugin supports a single sensor on the main line.  The sensor is monitored by an Arduino (or clone - I'm using an [Adafruit Metro Mini 328](https://www.adafruit.com/product/2590)).  The Pi pulls in the flow information from the Arduino through an I2C bus.  Given that the new recommendation is to drive the relays through I2C as well, I want to test to see if we can share the bus.  There's a possible logic level issue, though.  I want to see if I can make it work without the additional complication of a level shifting chip.

I like that there's the possibility to put the flow info on the main page. I'll take a look at the code. I've attached a couple of screen shots of the two plugin pages I've developed.  The program is largely functional now, but I still have a handful of coding tasks, QA, and documentation left to complete.  Then, I want to see it actually working in my garden before I share it with others. :)

I think you've done a great job with the plugin framework.  I've enjoyed working with it and have learned a lot from it.

Ian

![image](https://user-images.githubusercontent.com/20232907/167542839-e040e4e7-b403-4213-a47c-6da289a84e53.png)
![image](https://user-images.githubusercontent.com/20232907/167542867-cc8d3435-0e96-4293-b041-112defa27c08.png)

**Reply by Dan-in-CA** [reply]
Looks great!

It looks like you are using a separate log file for the flow data. That's a good thing. I have tinkered with having plugins add data to the run log but it has been kind of flaky so far.

On your flow settings page you have a volume measure setting. I am guessing that allows for SI (metric) units? Because SIP is used around the world I always try to have the option for either imperial or SI units to be used.

**Reply by IMHarris** [reply]
I did toy with the idea of adding to the main log, but decided against it for two reasons.  First, I didn't want to get in the way of future development of the main program.  Second, and this was the deal killer, the flow log ends up having to cut the data differently.  In the scenario where there are two valves on at the same time (see last log entry in screen grab above) the sensor, which is installed on the main line, cannot distinguish the flow between the valves.  So, the log needs to group that scenario into a single entry - unlike the main log which would report that out in two rows.

As for the metric measurements, the short answer is yes.

The flow sensors send an electric pulse per unit volume.  Because sensors can all be different and I can't know what people will use, I created a calibration mode for the program.  The mode will tell the user how many pulses it counted as they fill a container of a known size:

![image](https://user-images.githubusercontent.com/20232907/167682126-bb38f1c7-7c8c-4973-9924-01d7bd9fac7e.png)
![image](https://user-images.githubusercontent.com/20232907/167682159-55ad844f-e518-4c59-afd3-d2cb53f185ba.png)

So, the user could fill a five liter bucket, divide the count of pulses reported by five and then enter that number into the "Sensor pulses per volume" setting along with "liters" in the "Volume measure" setting

**Reply by Dan-in-CA** [reply]
Your calibration strategy makes the plugin quite versatile.
Looks like this will be a very useful plugin.

**Reply by IMHarris** [reply]
Thanks Dan.  I'm hoping so.  Will check back in as I get closer to deployment.
